src/Spectre.libGaussianMixtureModelling/SplitterSegmentExtractor.cpp

```diff
@@ -0,0 +1,224 @@
+    Index indexOfSplittingPeak, DataType resolutionCoeff, size_t peaksNum)
+{
+    Index rightmostValleyIndex = std::min((unsigned)peaksNum - 1,
+        indexOfSplittingPeak + CLEAR_PEAK_SEARCH_RANGE - 1);
+    Index intervalEnd = valleyIndices(rightmostValleyIndex) - 1;
+
+    DataType lastMz = spectrum.mzs[intervalEnd];
+    DataType splittingPeakMz = spectrum.mzs[indexOfSplittingPeak];
+
+    DataType minMzRange = CLEAR_PEAK_SEARCH_RANGE*splittingPeakMz*resolutionCoeff;
+
+    if (lastMz - splittingPeakMz < minMzRange)
+        ExtendIntervalToRight(intervalEnd, minMzRange, spectrum);
+
+    return intervalEnd;
+}
+
+static inline Index GetIntervalStart(SpectrumView spectrum, IndicesView valleyIndices,
+    Index indexOfSplittingPeak, DataType resolutionCoeff)
+{
+    Index leftmostValleyIndex =
+        std::max(0, ((int)indexOfSplittingPeak - CLEAR_PEAK_SEARCH_RANGE));
+    Index intervalStart = std::max((int)valleyIndices[leftmostValleyIndex] - 1, 0);
+
+    DataType startingMz = spectrum.mzs[intervalStart];
+    DataType splittingPeakMz = spectrum.mzs[indexOfSplittingPeak];
+
+    DataType minMzRange = CLEAR_PEAK_SEARCH_RANGE*splittingPeakMz*resolutionCoeff;
+    if (splittingPeakMz - startingMz < minMzRange)
+        ExtendIntervalToLeft(intervalStart, minMzRange, spectrum);
+
+    return intervalStart;
+}
+
+static inline Data GetMzsOfLeftMargin(SpectrumView& spectrum, Index intervalStart,
+    DataType resolutionCoeff)
+{
+    DataView& mzs = spectrum.mzs;
+    const DataType step = mzs[intervalStart + 1] - mzs[intervalStart];
+    DataType leftMarginMzBegin = mzs[intervalStart] * (1.0 - 6.0*resolutionCoeff);
+    DataType leftMarginMzEnd = mzs[intervalStart] - step;
+    size_t intervalSize =
+        (size_t)ceil((leftMarginMzEnd - leftMarginMzBegin) / step);
+
+    Data margin(intervalSize);
+    DataType floatIter = leftMarginMzBegin;
+    for (Index i = 0; i < intervalSize; i++, floatIter += step)
+    {
+        margin[i] = floatIter;
+    }
+
+    return margin;
+}
+
+static inline Data GetMzsOfRightMargin(SpectrumView& spectrum, Index intervalEnd,
+    DataType resolutionCoeff)
+{
+    DataView& mzs = spectrum.mzs;
+    const DataType step = mzs[intervalEnd] - mzs[intervalEnd - 1];
+    DataType marginMzBegin = mzs[intervalEnd] + step;
+    DataType marginMzEnd = mzs[intervalEnd] + 6 * resolutionCoeff*mzs[intervalEnd];
+    size_t intervalSize =
+        (size_t)ceil((marginMzEnd - marginMzBegin) / step);
+
+    Data margin(intervalSize);
+    DataType floatIter = marginMzBegin;
+    for (Index i = 0; i < intervalSize; i++, floatIter += step)
+    {
+        margin[i] = floatIter;
+    }
+
+    return margin;
+}
+
+static inline Data GetIntensitiesOfLeftMargin(SpectrumView spectrum, DataView margin,
+    Index intervalBegin, Index intervalEnd, DataType resolutionCoeff)
+{
+    const DataView& mzs = spectrum.mzs;
+
+    const DataType heightScaleFactor = sqrt(2.0 * M_PI) * 2.0 * resolutionCoeff *
```

Comment:
Constant `sqrt(2.0 * M_PI) * 2.0` as `constexpr` maybe?

Reply:
Any good naming ideas? I've already spent too much trying to come up with one that would make sense. Good compiler (not MSVC) should automatically detect this as constant and compute it in compile time.

Reply:
`heightScaleConstant`?

Reply:
Not all that informative is it? :smile: